engine/src/main/java/org/hibernate/validator/internal/engine/ValidatorImpl.java

```diff
@@ -156,7 +158,17 @@ public ValidatorImpl(ConstraintValidatorFactory constraintValidatorFactory,
 		this.validationOrderGenerator = validationOrderGenerator;
 		this.traversableResolver = traversableResolver;
 		this.validatorScopedContext = new ValidatorScopedContext( messageInterpolator, parameterNameProvider, clockProvider, temporalValidationTolerance, scriptEvaluatorFactory, failFast, traversableResolverResultCacheEnabled );
-		this.constraintValidatorInitializationContext = new HibernateConstraintValidatorInitializationContextImpl( validatorScopedContext.getScriptEvaluatorFactory(), validatorScopedContext.getClockProvider(), validatorScopedContext.getTemporalValidationTolerance() );
+		this.scriptEvaluatorFactory = validatorScopedContext.getScriptEvaluatorFactory();
+		this.constraintValidatorInitializationContext = new HibernateConstraintValidatorInitializationContextImpl( validatorScopedContext.getScriptEvaluatorFactory(), validatorScopedContext.getClockProvider(), validatorScopedContext.getTemporalValidationTolerance(), null );
+	}
+
+	public ValidatorImpl withDynamicConstraintValidatorInitializationPayload(Object dynamicPayload) {
+		this.constraintValidatorInitializationContext = new HibernateConstraintValidatorInitializationContextImpl(
```

Comment:
Concurrency issue. You're not overriding the context for your `validate()` call here but also for all the other random threads running. And the other threads will override it too.

I think we won't be able to make your approach work.

My opiniong is that you'd be better served by having the payload at the `ValidatorFactory` level AND the `HibernateValidatorContext` level.

The drawback is that doing you calls to `validate()` you would have to instantiate the `Validator` first by unwrapping the context and setting the payload. The `Validator` creation is quite lightweight so I don't think it will be an issue in practice.

Could you try this approach and see how it goes with your use case?